File: HackerRank/101hack24/C.cpp
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <cctype>
#include <cmath>
#include <ctime>
#include <cassert>
#include <climits>
#include <iostream>
#include <iomanip>
#include <string>
#include <vector>
#include <set>
#include <map>
#include <queue>
#include <deque>
#include <bitset>
#include <algorithm>

#define MST(a, b) memset(a, b, sizeof(a))
#define REP(i, a) for (int i = 0; i < int(a); i++)
#define REPP(i, a, b) for (int i = int(a); i <= int(b); i++)
#define LL long long

using namespace std;

const int N = 3005;
int step[N], cost[N], n, talk;

bool check(LL tot) {
    REPP(i, 1, n) step[i] = 0;
    int person = 2;
    REPP(i, 1, n) if (i == 1 || step[i]) {
        LL turn = (tot - 1LL * step[i] * talk - cost[i]) / talk;
        int now = step[i];
        int lim = min(1LL * n, turn);
        REP(j, lim) {
            step[person++] = ++now;
            if (person > n) break;
        }
        step[i] = now;
        if (person > n) break;
    }
    REPP(i, 1, n) if ((step[i] == 0 && i > 1) || 1LL * step[i] * talk + cost[i] > tot) return 0;
    return 1;
}

int main() {
    ios :: sync_with_stdio(0);
    //freopen("C.in", "r", stdin);
    int t;
    cin >> t;
    while (t--) {
        cin >> n >> talk;
        REPP(i, 1, n) cin >> cost[i];
        sort(cost + 1, cost + n + 1);
        reverse(cost + 1, cost + n + 1);
        LL lo = 0, hi = 1e15;
        while (lo + 1 < hi) {
            LL mid = (lo + hi) >> 1;
            if (check(mid)) hi = mid;
            else lo = mid;
        }
        cout << hi << endl;
    }

    return 0;
}

File: HackerRank/101hack24/C.py
cost = []

def check(tot):
    step = [0] * n
    person = 1
    for i in range(n):
        if (i == 0 or step[i]):
            if person >= n:
                break
            turn = (tot - step[i] * talk - cost[i]) / talk
            now = step[i]; limit = min(n, turn)
            for j in range(limit):
                now += 1
                step[person] = now
                person += 1
                if person >= n:
                    break
            step[i] = now
    for i in range(n):
        if (step[i] == 0 and i > 0) or step[i] * talk + cost[i] > tot:
            return False
    return True

T = int(raw_input())
for go in range(T):
    n, talk = map(int, raw_input().split())
    cost = map(int, raw_input().split())
    cost.sort(); cost.reverse()
    lo = 0; hi = int(1e15)
    while lo + 1 < hi:
        mid = (lo + hi) / 2
        if check(mid):
            hi = mid
        else :
            lo = mid
    print hi

                
                
